contrib/babelfishpg_tsql/src/pl_exec-2.c

```diff
@@ -3112,6 +3119,297 @@ exec_stmt_insert_bulk(PLtsql_execstate *estate, PLtsql_stmt_insert_bulk *stmt)
+		pfree(db_name);
+	}
+
+	login_is_db_owner = 0 == strncmp(GetUserNameFromId(GetSessionUserId(), false),
+										get_owner_of_db(db_name), NAMEDATALEN);
+	user = get_user_for_database(db_name);
+	guest_role_name = get_guest_role_name(db_name);
+
+    if(is_cross_db)
+    {
+        if (user)
+            SetCurrentRoleId(GetSessionUserId(), false);
+        else
+            ereport(ERROR,
+                    (errcode(ERRCODE_UNDEFINED_DATABASE),
+                     errmsg("The server principal \"%s\" is not able to access "
+                            "the database \"%s\" under the current security context",
+                            login, dbcc_stmt.db_name)));
+    }
+    pfree(login);
+
+	if (dbcc_stmt.schema_name)
+	{
+		schema_name = pstrdup(dbcc_stmt.schema_name);
+		nsp_name = get_physical_schema_name(db_name, schema_name);
+	}
+	else
+	{
+		/* 
+		 * If schema_name is not provided, find default schema for current user
+		 * and get physical schema name
+		 */
+		if (!user)
+		{
+			ereport(ERROR,
+			(errcode(ERRCODE_UNDEFINED_OBJECT),
+				 errmsg("user does not exist")));
+		}
+		else if ((guest_role_name && strcmp(user, guest_role_name) == 0))
+		{
+			nsp_name = pstrdup(get_guest_schema_name(db_name));
+		}
+		else
+		{
+			schema_name = get_authid_user_ext_schema_name((const char *) db_name, user);
+			nsp_name = get_physical_schema_name(db_name, schema_name);
+		}
+	}
+
+	pfree(db_name);
+
+	nsp_oid = get_namespace_oid(nsp_name, false);
+	if(!OidIsValid(nsp_oid))
+	{
+		ereport(ERROR,
+		(errcode(ERRCODE_UNDEFINED_SCHEMA),
+			errmsg("schema \"%s\" does not exist", schema_name)));
+	}
+
+	/* Permission check */
+	if (!(pg_namespace_ownercheck(nsp_oid, GetUserId()) ||
+			has_privs_of_role(GetSessionUserId(), get_role_oid("sysadmin", false)) ||
+				login_is_db_owner))
+		aclcheck_error(ACLCHECK_NOT_OWNER, OBJECT_SCHEMA, nsp_name);
+
+	table_name = pstrdup(dbcc_stmt.table_name);
+	table_oid = get_relname_relid(table_name, nsp_oid);
+	if(!OidIsValid(table_oid))
+	{
+		ereport(ERROR,
+		(errcode(ERRCODE_UNDEFINED_TABLE),
+			errmsg("relation \"%s\" does not exist", table_name)));
+	}
+
+	rel = RelationIdGetRelation(table_oid);
+	tupdesc = RelationGetDescr(rel);
+
+	/* Find Identity column in table and associated sequence */
+	for (attnum = 0; attnum < tupdesc->natts; attnum++)
+	{
+		Form_pg_attribute attr = TupleDescAttr(tupdesc, attnum);
+
+		if (attr->attidentity)
+		{
+			attname = NameStr(attr->attname);
+			seqid = getIdentitySequence(table_oid, attnum + 1, false);
+			break;
+		}
+	}
+
+	RelationClose(rel);
+
+	if (!OidIsValid(seqid))
+		ereport(ERROR,
+		(errcode(ERRCODE_UNDEFINED_COLUMN),
+			errmsg("'%s.%s' does not contain an identity column.",
+				nsp_name, table_name)));
+
+	PG_TRY();
+	{
+		// Lock table to ensure concurrency.
+		if(dbcc_stmt.is_reseed)
+		{
+			LockRelationOid( table_oid, AccessExclusiveLock);
+		}
+		else
+		{
+			LockRelationOid( table_oid, ShareLock);
```

Comment:
Please add a concurrency test for this scenario where other backends are blocked from accessing this table while a DBCC CHECKIDENT has reseeded the identity column. You can use isolation test framework.

Reply:
added test for concurrency using isolation test framework.